arangod/Aql/IResearchViewOptimizerRules.cpp

```diff
@@ -305,74 +296,168 @@ bool optimizeScoreSort(IResearchViewNode& viewNode, ExecutionPlan* plan) {
+                              std::numeric_limits<ptrdiff_t>::max() &&
+                          nameFromAttributeAccess(name, *astCalcNode, ctx,
+                                                  false)) {
+                        element.source = column.first;
+                        element.fieldNumber = column.second;
+                        element.postfix = name;
+                        element.ascending = sort.ascending;
+                      }
+                    }
+                  });
+              if (usedVars != 1 || element.isScore()) {
+                return false;
+              }
+              heapSort.push_back(std::move(element));
+            }
+            break;
+          default:
+            return false;
+        }
+      } break;
       default:
         TRI_ASSERT(false);
         return false;
     }
-
-    TRI_ASSERT(sortVariable);
-    auto const s = std::find_if(
-        std::begin(scorers), std::end(scorers),
-        [sortVariableId = sortVariable->id](auto const& t) noexcept {
-          return t.var->id == sortVariableId;
-        });
-    if (s == std::end(scorers)) {
+  }
+  if (!attrs.empty()) {
+    if (latematerialized::attributesMatch<true>(
+            primarySort, storedValues, attrs, usedColumns, columnsCount)) {
+      latematerialized::setAttributesMaxMatchedColumns<true>(usedColumns,
+                                                             columnsCount);
+      auto const attrCount = attrs.size();
+      for (size_t i = 0; i < attrCount; ++i) {
+        auto const& a = attrs[i];
+        TRI_ASSERT(storedMaps.contains(i));
+        auto& sortBucket = heapSort[storedMaps[i]];
+        sortBucket.source = a.afData.columnNumber;
+        sortBucket.fieldNumber = a.afData.fieldNumber;
+        TRI_ASSERT(sortBucket.postfix.empty());
+        TRI_ASSERT(a.afData.field);
+        auto const fieldSize = a.afData.field->size();
+        TRI_ASSERT(fieldSize > a.afData.postfix);
+        for (size_t i = a.afData.postfix + 1; i < fieldSize; ++i) {
+          if (i != a.afData.postfix + 1) {
+            sortBucket.postfix += ".";
+          }
+          sortBucket.postfix += a.afData.field->at(i).name;
+        }
+      }
+    } else {
       return false;
     }
-    auto pos = std::distance(scorers.begin(), s);
-    if (pos == 0) {
-      zero = scoresSort.size();
-    }
-    scoresSort.emplace_back(pos, sort.ascending);
   }
-
-  auto& idx = scoresSort.front().first;
-  if (idx != 0) {
+  if (heapSort.front().isScore() && heapSort.front().source != 0) {
+    auto idx = heapSort.front().source;
     std::swap(scorers.front(), scorers[idx]);
-    idx = 0;
-    if (zero != std::numeric_limits<size_t>::max()) {
-      scoresSort[zero].first = idx;
+    for (auto it = heapSort.begin(); it != heapSort.end(); ++it) {
+      if (it->isScore() && it->source == 0) {
+        it->source = idx;
+      } else if (it->isScore() && it->source == idx) {
+        it->source = 0;
+      }
     }
```

Comment:
Please keep old code, it was faster and simpler

Reply:
Yes, but it didn't covered the case when one score used several times in sort. Not very common case  but still  valid AQL.

Reply:
Hmm, case like `let x = TFIDF(d) SORT BM25(d), x, BM25(d)`?
BM25 will be deduplicated?

Reply:
No.

Reply:
SORT BM25(d), TFIDF(d), BM25(d)  DESC  -> SORT #11 ASC, #9 ASC, #11 DESC

Reply:
It looks like yes

Reply:
I meant compute will be deduplicated